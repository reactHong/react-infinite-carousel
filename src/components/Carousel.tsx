import React, { useEffect, useRef, useState } from 'react';
import CarouselHeader from './CarouselHeader';
import CarouselTrack, { Direction } from './CarouselTrack';
import { Item, CardInfo } from './CarouselCard';

type CarouselProps = {
  title: string;
  propsItems: Item[];
  maxCarouselCardNum: number;
  enableContinuousClick: boolean;
};

type CarouselState = {
  carouselRange: CarouselRange;
  items: Item[];
  renderItems: Item[];
  translateX: number;
  direction: Direction;
};

type CarouselRange = {
  first: number;
  last: number;
};

//TODO: EMPTY_ITEM can be changed to copiedItem except the real image URL
export const EMPTY_ITEM: Item = {
  blurhash: "UAN=8k?LS~M:ErJFs%t0MDMWRqo@%BxSV{RX",
  launch_date: "",
  location: [],
  name: "emptyCard",
  online: false,
  popularity: 0,
};

//TODO: Consider cross browsing
const getElementTranslateX = (element: HTMLDivElement): number => {
  const style: CSSStyleDeclaration = window.getComputedStyle(element);
  const values = style.transform.split(/\w+\(|\);?/);
  const transform = values[1].split(/,\s?/g).map(numStr => parseInt(numStr));
  return transform[0];
};

const reloadCardInfo = (carouselWidth: number, maxCarouselCardNum: number): CardInfo => {
  //TODO: This depends on this.state.carouselWidth에 depend on. 
  //      cardInfo should be reset if the width is changed.
  //      Think about good structure when resizing the window
  const cardGap: number = 10;
  const hvRatio: number = 9 / 16;
  const cardWidth: number = (carouselWidth - (cardGap * (maxCarouselCardNum - 1))) / maxCarouselCardNum;
  const cardHeight: number = cardWidth * hvRatio;

  return {
    width: cardWidth,
    imageHeight: cardHeight,
    gap: cardGap,
  };
};

const carouselChecker = (items: Item[], range: CarouselRange) => {
  const cItems = [...items];
  const cRange = { ...range };

  const isPrevItemEmpty = () => ((cRange.first === 0) ? true : false);
  const isNextItemEmpty = () => ((cRange.last >= cItems.length - 1) ? true : false);

  const prev = (): boolean => {
    let shouldReorder: boolean = false;
    if (isPrevItemEmpty()) {
      cItems.unshift(cItems.pop()!);
      shouldReorder = true;
    } else {
      cRange.first--;
      cRange.last--;
    }
    return shouldReorder;
  };

  const next = (): boolean => {
    let shouldReorder: boolean = false;
    if (isNextItemEmpty()) {
      cItems.push(cItems.shift()!)
      shouldReorder = true;
    } else {
      cRange.first++;
      cRange.last++;
    }
    return shouldReorder;
  };

  return {
    items: cItems,
    range: cRange,
    prev: prev,
    next: next,
  }
};

function Carousel({
  title,
  propsItems,
  maxCarouselCardNum,
  enableContinuousClick,
}: CarouselProps) {

  //TODO: It should be guaranted that propsItems is always not empty
  // if (propsItems.length <= 0) {
  //   //TODO: Error Handling
  //   return <div></div>;
  // }

  const [cardInfo, _setCardInfo] = useState<CardInfo>({
    width: 0,
    imageHeight: 0,
    gap: 0,
  });
  const [state, _setState] = useState<CarouselState>({
    carouselRange: {  // For checking if the prev/next item exists
      first: 0,
      last: maxCarouselCardNum - 1,
    },
    items: [...propsItems],
    renderItems: [...propsItems],
    translateX: 0,
    direction: 'none',
  });

  //NOTE: stateRef is needed for using state in eventlisenter - Use it in transitionEnd
  const cardInfoRef = React.useRef<CardInfo>(cardInfo);
  const stateRef = React.useRef<CarouselState>(state);
  const setCardInfo = (newCardInfo: CardInfo) => {
    cardInfoRef.current = newCardInfo;
    _setCardInfo(newCardInfo);
  };
  const setState = (newState: CarouselState) => {
    stateRef.current = newState;
    _setState(newState);
  };

  const trackContainerRef = useRef<HTMLDivElement>(null);
  const cardTrackRef = useRef<HTMLDivElement>(null);

  const handlePrevClick = () => {
    if (enableContinuousClick && state.direction === "next") return;
    if (!enableContinuousClick && state.direction !== "none") return;

    const checker = carouselChecker(state.items, state.carouselRange);
    const shouldReorder = checker.prev();
    const renderItems: Item[] = [...checker.items];

    const translateXUnit: number = cardInfo.width + cardInfo.gap;
    console.log("prev: translateXUnit:", translateXUnit);
    let newDirection: Direction = "prev";
    let newTranslateX: number = state.translateX + translateXUnit;

    if (shouldReorder) {
      if (!cardTrackRef.current) {
        //TODO: Error handling - It means something UI could be changed
        return;
      }

      //TODO: "emptyCard" should be compared to id
      const emptyItems = state.renderItems.filter(item => item.name === EMPTY_ITEM.name);
      emptyItems.push(EMPTY_ITEM);
      renderItems.push(...emptyItems);

      //NOTE: With new direction, transition would be none in a while for repositioning.
      const currentTranslateX = getElementTranslateX(cardTrackRef.current);
      newTranslateX = currentTranslateX - translateXUnit;
      newDirection = "none";

      setTimeout(() => {
        const state = stateRef.current;
        setState({
          ...state,
          translateX: 0,
          direction: "prev",
        });
      }, 20);
    }

    setState({
      ...state,
      carouselRange: { ...checker.range },
      items: [...checker.items],
      renderItems: renderItems,
      translateX: newTranslateX,
      direction: newDirection,
    });
  };

  const handleNextClick = () => {
    if (enableContinuousClick && state.direction === "prev") return;
    if (!enableContinuousClick && state.direction !== "none") return;

    const checker = carouselChecker(state.items, state.carouselRange);
    const shouldReorder = checker.next();
    const translateXUnit: number = cardInfo.width + cardInfo.gap;
    console.log("next: translateXUnit:", translateXUnit);
    const translateX: number = state.translateX - translateXUnit;
    const renderItems: Item[] = [...checker.items];

    if (shouldReorder) {
      //TODO: "emptyCard" should be compared to id
      const emptyItems: Item[] = state.renderItems.filter(item => item.name === EMPTY_ITEM.name);
      emptyItems.unshift(EMPTY_ITEM);
      renderItems.unshift(...emptyItems);
    }

    setState({
      ...state,
      items: [...checker.items],
      carouselRange: { ...checker.range },
      renderItems: renderItems,
      translateX: translateX,
      direction: "next",
    });
  };

  const transitionEnd = () => {
    const state = stateRef.current;
    const cardInfo = cardInfoRef.current;

    let emptyCardsCount: number = 0;
    const newRenderItems: Item[] = state.renderItems.filter(item => {
      //TODO: "emptyCard" should be compared to id
      if (item.name === EMPTY_ITEM.name) {
        emptyCardsCount++;
        return false;
      }
      return true;
    });
    const newState: CarouselState = {
      ...state,
      renderItems: newRenderItems,
      direction: "none",
    };

    if (state.direction === "next") {
      const translateXUnit = cardInfo.width + cardInfo.gap;
      const newTranslateX = state.translateX + translateXUnit * emptyCardsCount;
      newState.translateX = newTranslateX;
    }
    setState(newState);
  };

  const handleResize = () => {
    if (!trackContainerRef.current) return;

    const carouselWidth: number = trackContainerRef.current!.clientWidth;
    const cardInfo = reloadCardInfo(carouselWidth, maxCarouselCardNum);
    setCardInfo(cardInfo);
  };

  // console.log("\n[Carousel] state:", state);
  useEffect(() => {
    // console.log("[Carousel.useEffect]");

    //TODO: Render 3 times when first loading 
    //  1. for getting carouselWidth - trackContainerRef.current!.clientWidth
    //  2. for rendering loading screen
    //  3. for rendering with actual data
    if (!trackContainerRef.current) return;

    const carouselWidth: number = trackContainerRef.current!.clientWidth;
    const cardInfo = reloadCardInfo(carouselWidth, maxCarouselCardNum);
    setCardInfo(cardInfo);

    // console.log("[Carousel.useEffect] cardTrackRef.current:", cardTrackRef.current);
    window.addEventListener("resize", handleResize);
    cardTrackRef.current?.addEventListener("transitionend", transitionEnd);
    return () => {
      console.log("removeEventListener");
      cardTrackRef.current?.removeEventListener("transitionend", transitionEnd);
      window.removeEventListener("resize", handleResize);
    };
  }, []);

  // console.log("-- [Carousel.render]!!!!");
  return (
    <div className="carouselContainer">
      <CarouselHeader title={title} prev={handlePrevClick} next={handleNextClick} />
      <CarouselTrack
        trackContainerRef={trackContainerRef}
        cardTrackRef={cardTrackRef}
        cardInfo={cardInfo}
        direction={state.direction}
        translateX={state.translateX}
        renderItems={state.renderItems}
      />
    </div>
  );
}

export default Carousel;