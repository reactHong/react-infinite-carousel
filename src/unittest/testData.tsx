export const testDiscoveryData = {
  sections: [{
    title: 'New Restaurants',
    restaurants: [
      {
        blurhash: "UAN=8k?LS~M:ErJFs%t0MDMWRqo@%BxSV{RX",
        launch_date: "",
        location: [],
        name: "0",
        online: false,
        popularity: 0,
      }, {
        blurhash: "UMD*~9$$azNK9}I@jts-12xEoKaz=wodS3R,",
        launch_date: "",
        location: [],
        name: "1",
        online: false,
        popularity: 0,
      }, {
        blurhash: "U6T69eube;rHy,lNf%a3V5aOkUknu:ZTg1k-",
        launch_date: "",
        location: [],
        name: "2",
        online: false,
        popularity: 0,
      }, {
        blurhash: "UCO;.s:bO%r_yWXlORbbC?TFvobaVDi_t9nS",
        launch_date: "",
        location: [],
        name: "3",
        online: false,
        popularity: 0,
      }, {
        blurhash: "U9O[r*?hI_VN*8yNniVx5^NhxTknY]MmX+tx",

        launch_date: "",
        location: [],
        name: "4",
        online: false,
        popularity: 0,
      }, {
        blurhash: "UQCBXU,@JQn,${xEWXWo31OTwzW-N4Ncs+sV",
        launch_date: "",
        location: [],
        name: "5",
        online: false,
        popularity: 0,
      },
    ]
  }]
};
