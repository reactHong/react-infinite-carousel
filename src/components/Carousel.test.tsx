import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Carousel from './Carousel';
import { testDiscoveryData } from '../testData';

const MAX_NUM = 5;
const continuousClick = true;
const section = testDiscoveryData.sections[0];

describe('<Carousel />', () => {
  it('matches snapshot', () => {
    const utils = render(<Carousel
      key={0}
      title={section.title}
      propsItems={section.restaurants}
      maxCarouselCardNum={MAX_NUM}
      enableContinuousClick={continuousClick}
    />);
    expect(utils.container).toMatchSnapshot();
  });



});