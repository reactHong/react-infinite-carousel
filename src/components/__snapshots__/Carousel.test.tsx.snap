// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> matches snapshot 1`] = `
<div>
  <div
    class="carouselContainer"
  >
    <div
      class="headerContainer"
    >
      <h2
        class="header"
      >
        New Restaurants
      </h2>
      <div
        class="buttonsContainer"
      >
        <button
          type="button"
        >
          <svg
            class="ic_m_caret_left_svg__rtlIcon"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M15.477 3.977a.503.503 0 01.365.841l-6.68 7.159 6.679 7.159a.502.502 0 01.159.366.5.5 0 01-.888.315l-7-7.5a.498.498 0 010-.682l7-7.5a.496.496 0 01.365-.16"
              fill="currentColor"
            />
          </svg>
        </button>
        <button
          type="button"
        >
          <svg
            class="ic_m_caret_right_svg__rtlIcon"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M8.5 20a.5.5 0 01-.365-.841L14.815 12 8.136 4.841a.5.5 0 11.729-.681l7 7.5c.18.192.18.49 0 .682l-7 7.5a.494.494 0 01-.365.16"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="trackContainer"
    >
      <div
        class="cardTrack"
        style="transform: translate3d(0px, 0px, 0px); transition: none;"
      >
        <div
          class="card"
          style="width: -8px; min-width: -8px; margin-right: 10px;"
        >
          <div
            class="top"
            style="height: -4.5px;"
          >
            <div
              class="imageContainer"
            >
              <div
                style="display: inline-block; height: 100%; width: 100%; position: relative;"
              >
                <canvas
                  height="32"
                  style="position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; width: 100%; height: 100%;"
                  width="32"
                />
              </div>
            </div>
            <div
              class="online"
            >
              <div>
                offline
              </div>
            </div>
          </div>
          <div
            class="bottom"
          >
            <div>
              0
            </div>
          </div>
        </div>
        <div
          class="card"
          style="width: -8px; min-width: -8px; margin-right: 10px;"
        >
          <div
            class="top"
            style="height: -4.5px;"
          >
            <div
              class="imageContainer"
            >
              <div
                style="display: inline-block; height: 100%; width: 100%; position: relative;"
              >
                <canvas
                  height="32"
                  style="position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; width: 100%; height: 100%;"
                  width="32"
                />
              </div>
            </div>
            <div
              class="online"
            >
              <div>
                offline
              </div>
            </div>
          </div>
          <div
            class="bottom"
          >
            <div>
              1
            </div>
          </div>
        </div>
        <div
          class="card"
          style="width: -8px; min-width: -8px; margin-right: 10px;"
        >
          <div
            class="top"
            style="height: -4.5px;"
          >
            <div
              class="imageContainer"
            >
              <div
                style="display: inline-block; height: 100%; width: 100%; position: relative;"
              >
                <canvas
                  height="32"
                  style="position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; width: 100%; height: 100%;"
                  width="32"
                />
              </div>
            </div>
            <div
              class="online"
            >
              <div>
                offline
              </div>
            </div>
          </div>
          <div
            class="bottom"
          >
            <div>
              2
            </div>
          </div>
        </div>
        <div
          class="card"
          style="width: -8px; min-width: -8px; margin-right: 10px;"
        >
          <div
            class="top"
            style="height: -4.5px;"
          >
            <div
              class="imageContainer"
            >
              <div
                style="display: inline-block; height: 100%; width: 100%; position: relative;"
              >
                <canvas
                  height="32"
                  style="position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; width: 100%; height: 100%;"
                  width="32"
                />
              </div>
            </div>
            <div
              class="online"
            >
              <div>
                offline
              </div>
            </div>
          </div>
          <div
            class="bottom"
          >
            <div>
              3
            </div>
          </div>
        </div>
        <div
          class="card"
          style="width: -8px; min-width: -8px; margin-right: 10px;"
        >
          <div
            class="top"
            style="height: -4.5px;"
          >
            <div
              class="imageContainer"
            >
              <div
                style="display: inline-block; height: 100%; width: 100%; position: relative;"
              >
                <canvas
                  height="32"
                  style="position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; width: 100%; height: 100%;"
                  width="32"
                />
              </div>
            </div>
            <div
              class="online"
            >
              <div>
                offline
              </div>
            </div>
          </div>
          <div
            class="bottom"
          >
            <div>
              4
            </div>
          </div>
        </div>
        <div
          class="card"
          style="width: -8px; min-width: -8px; margin-right: 10px;"
        >
          <div
            class="top"
            style="height: -4.5px;"
          >
            <div
              class="imageContainer"
            >
              <div
                style="display: inline-block; height: 100%; width: 100%; position: relative;"
              >
                <canvas
                  height="32"
                  style="position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; width: 100%; height: 100%;"
                  width="32"
                />
              </div>
            </div>
            <div
              class="online"
            >
              <div>
                offline
              </div>
            </div>
          </div>
          <div
            class="bottom"
          >
            <div>
              5
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
